File: docs/breez-sdk/snippets/flutter/lib/htlcs.dart
import 'package:convert/convert.dart';
import 'package:breez_sdk_spark_flutter/breez_sdk_spark.dart';
import 'package:crypto/crypto.dart';

Future<Payment> sendHtlcPayment(BreezSdk sdk) async {
  // ANCHOR: send-htlc-payment
  String paymentRequest = "<spark address>";
  // Set the amount you wish the pay the receiver
  BigInt? amountSats = BigInt.from(50000);
  final prepareRequest = PrepareSendPaymentRequest(
      paymentRequest: paymentRequest,
      amount: amountSats,
      tokenIdentifier: null,
      conversionOptions: null,
      feePolicy: null);
  final prepareResponse = await sdk.prepareSendPayment(request: prepareRequest);

  // If the fees are acceptable, continue to create the HTLC Payment
  final paymentMethod = prepareResponse.paymentMethod;
  if (paymentMethod is SendPaymentMethod_SparkAddress) {
    final fee = paymentMethod.fee;
    print("Fees: $fee sats");
  }

  String preimage = "<32-byte unique preimage hex>";
  List<int> preimageBytes = hex.decode(preimage);
  Digest paymentHashDigest = sha256.convert(preimageBytes);
  String paymentHash = hex.encode(paymentHashDigest.bytes);

  // Set the HTLC options
  final htlcOptions = SparkHtlcOptions(
      paymentHash: paymentHash, expiryDurationSecs: BigInt.from(1000));
  final options = SendPaymentOptions.sparkAddress(htlcOptions: htlcOptions);

  final request =
      SendPaymentRequest(prepareResponse: prepareResponse, options: options);
  final sendResponse = await sdk.sendPayment(request: request);
  final payment = sendResponse.payment;
  // ANCHOR_END: send-htlc-payment
  return payment;
}

Future<void> receiveHodlInvoicePayment(BreezSdk sdk) async {
  // ANCHOR: receive-hodl-invoice-payment
  String preimage = "<32-byte unique preimage hex>";
  List<int> preimageBytes = hex.decode(preimage);
  Digest paymentHashDigest = sha256.convert(preimageBytes);
  String paymentHash = hex.encode(paymentHashDigest.bytes);

  final response = await sdk.receivePayment(
      request: ReceivePaymentRequest(
          paymentMethod: ReceivePaymentMethod.bolt11Invoice(
              description: "HODL invoice",
              amountSats: BigInt.from(50000),
              expirySecs: null,
              paymentHash: paymentHash)));

  final invoice = response.paymentRequest;
  print("HODL invoice: $invoice");
  // ANCHOR_END: receive-hodl-invoice-payment
}

Future<List<Payment>> listClaimableHtlcPayments(BreezSdk sdk) async {
  // ANCHOR: list-claimable-htlc-payments
  final request = ListPaymentsRequest(
    typeFilter: [PaymentType.receive],
    statusFilter: [PaymentStatus.pending],
    paymentDetailsFilter: [
      PaymentDetailsFilter.spark(
        htlcStatus: [SparkHtlcStatus.waitingForPreimage],
      ),
      PaymentDetailsFilter.lightning(
        htlcStatus: [SparkHtlcStatus.waitingForPreimage],
      ),
    ],
  );

  final response = await sdk.listPayments(request: request);
  final payments = response.payments;

  for (final payment in payments) {
    final details = payment.details;
    if (details is PaymentDetails_Spark && details.htlcDetails != null) {
      print("Spark HTLC expiry time: ${details.htlcDetails!.expiryTime}");
    } else if (details is PaymentDetails_Lightning) {
      print("Lightning HTLC expiry time: ${details.htlcDetails.expiryTime}");
    }
  }
  // ANCHOR_END: list-claimable-htlc-payments
  return payments;
}

Future<Payment> claimHtlcPayment(BreezSdk sdk) async {
  // ANCHOR: claim-htlc-payment
  String preimage = "<preimage hex>";
  final response = await sdk.claimHtlcPayment(
      request: ClaimHtlcPaymentRequest(preimage: preimage));
  final payment = response.payment;
  // ANCHOR_END: claim-htlc-payment
  return payment;
}

File: docs/breez-sdk/snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/Htlcs.kt
package com.example.kotlinmpplib

import breez_sdk_spark.*
import com.ionspin.kotlin.bignum.integer.BigInteger
import org.kotlincrypto.hash.sha2.SHA256

@OptIn(kotlin.ExperimentalStdlibApi::class)
class Htlcs {
    suspend fun sendHtlcPayment(sdk: BreezSdk) {
        // ANCHOR: send-htlc-payment
        val paymentRequest = "<spark address>"
        // Set the amount you wish the pay the receiver
        // Kotlin MPP (BigInteger from com.ionspin.kotlin.bignum.integer)
        val amountSats = BigInteger.fromLong(50_000L)
        // Android (BigInteger from java.math)
        // val amountSats = BigInteger.valueOf(50_000L)
        try {
            val prepareRequest = PrepareSendPaymentRequest(
                paymentRequest = paymentRequest,
                amount = amountSats,
                tokenIdentifier = null,
                conversionOptions = null,
                feePolicy = null,
            )
            val prepareResponse = sdk.prepareSendPayment(prepareRequest)

            // If the fees are acceptable, continue to create the HTLC Payment
            val paymentMethod = prepareResponse.paymentMethod
            if (paymentMethod is SendPaymentMethod.SparkAddress) {
                val fee = paymentMethod.fee
                // Log.v("Breez", "Fees: ${fee} sats")
            }

            val preimage = "<32-byte unique preimage hex>"
            val preimageBytes = preimage.hexToByteArray()
            val digest = SHA256()
            digest.update(preimageBytes)
            val paymentHashBytes = digest.digest()
            val paymentHash = paymentHashBytes.toHexString()

            // Set the HTLC options
            val htlcOptions = SparkHtlcOptions(
                paymentHash = paymentHash,
                expiryDurationSecs = 1000u
            )
            val options = SendPaymentOptions.SparkAddress(htlcOptions = htlcOptions)

            val request = SendPaymentRequest(
                prepareResponse = prepareResponse,
                options = options
            )
            val sendResponse = sdk.sendPayment(request)
            val payment = sendResponse.payment
        } catch (e: Exception) {
            // handle error
            throw e
        }
        // ANCHOR_END: send-htlc-payment
    }

    suspend fun receiveHodlInvoicePayment(sdk: BreezSdk) {
        // ANCHOR: receive-hodl-invoice-payment
        try {
            val preimage = "<32-byte unique preimage hex>"
            val preimageBytes = preimage.hexToByteArray()
            val digest = SHA256()
            digest.update(preimageBytes)
            val paymentHashBytes = digest.digest()
            val paymentHash = paymentHashBytes.toHexString()

            val response = sdk.receivePayment(
                ReceivePaymentRequest(
                    paymentMethod = ReceivePaymentMethod.Bolt11Invoice(
                        description = "HODL invoice",
                        amountSats = 50_000u,
                        expirySecs = null,
                        paymentHash = paymentHash
                    )
                )
            )

            val invoice = response.paymentRequest
            // Log.v("Breez", "HODL invoice: $invoice")
        } catch (e: Exception) {
            // handle error
            throw e
        }
        // ANCHOR_END: receive-hodl-invoice-payment
    }

    suspend fun listClaimableHtlcPayments(sdk: BreezSdk) {
        // ANCHOR: list-claimable-htlc-payments
        try {
            val request = ListPaymentsRequest(
                typeFilter = listOf(PaymentType.RECEIVE),
                statusFilter = listOf(PaymentStatus.PENDING),
                paymentDetailsFilter = listOf(
                    PaymentDetailsFilter.Spark(
                        htlcStatus = listOf(SparkHtlcStatus.WAITING_FOR_PREIMAGE),
                        conversionRefundNeeded = null
                    ),
                    PaymentDetailsFilter.Lightning(
                        htlcStatus = listOf(SparkHtlcStatus.WAITING_FOR_PREIMAGE)
                    )
                )
            )

            val response = sdk.listPayments(request)
            val payments = response.payments

            for (payment in payments) {
                val details = payment.details
                when (details) {
                    is PaymentDetails.Spark -> {
                        val htlc = details.htlcDetails
                        if (htlc != null) {
                            // Log.v("Breez", "Spark HTLC expiry time: ${htlc.expiryTime}")
                        }
                    }
                    is PaymentDetails.Lightning -> {
                        val htlc = details.htlcDetails
                        // Log.v("Breez", "Lightning HTLC expiry time: ${htlc.expiryTime}")
                    }
                    else -> {}
                }
            }
        } catch (e: Exception) {
            // handle error
            throw e
        }
        // ANCHOR_END: list-claimable-htlc-payments
    }

    suspend fun claimHtlcPayment(sdk: BreezSdk) {
        // ANCHOR: claim-htlc-payment
        try {
            val preimage = "<preimage hex>"
            val request = ClaimHtlcPaymentRequest(preimage = preimage)
            val response = sdk.claimHtlcPayment(request)
            val payment = response.payment
        } catch (e: Exception) {
            // handle error
            throw e
        }
        // ANCHOR_END: claim-htlc-payment
    }
}
